File: Contrib/SecureBlackbox.NET/Samples/VB.NET/MIMEBlackbox/Viewer/MimeViewer_Options.vb
Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports System.Threading

Imports SBMIME
Imports SBChSConv

Public Class MimeViewer_Options
    Inherits MimeViewer_PlugControl
    Private groupBox_parser As System.Windows.Forms.GroupBox
    Private lblHeaderCharset As System.Windows.Forms.Label
    Private WithEvents cmbHeaderCharset As System.Windows.Forms.ComboBox
    Private lblBodyCharset As System.Windows.Forms.Label
    Private WithEvents cmbBodyCharset As System.Windows.Forms.ComboBox
    Private WithEvents cmbActivatePartHandlers As System.Windows.Forms.CheckBox
    Private WithEvents chk_UseBackgroundParser As System.Windows.Forms.CheckBox
    Private groupBox_view As System.Windows.Forms.GroupBox
    Private WithEvents chkBodyInTree As System.Windows.Forms.CheckBox
    Private WithEvents chkHeaderInTree As System.Windows.Forms.CheckBox
    Private WithEvents chkFieldsInTree As System.Windows.Forms.CheckBox
    Private panel1 As System.Windows.Forms.Panel
    Private components As System.ComponentModel.IContainer = Nothing


    Public Sub New()
        ' This call is required by the Windows Form Designer.
        InitializeComponent()
    End Sub 'New

    ' TODO: Add any initialization after the InitializeComponent call
    ' Clean up any resources being used.
    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
        If disposing Then
            If Not (components Is Nothing) Then
                components.Dispose()
            End If
        End If
        MyBase.Dispose(disposing)
    End Sub 'Dispose

#Region "Designer generated code"

    ' Required method for Designer support - do not modify
    ' the contents of this method with the code editor.
    Private Sub InitializeComponent()
        Me.groupBox_parser = New System.Windows.Forms.GroupBox
        Me.chk_UseBackgroundParser = New System.Windows.Forms.CheckBox
        Me.cmbActivatePartHandlers = New System.Windows.Forms.CheckBox
        Me.cmbBodyCharset = New System.Windows.Forms.ComboBox
        Me.lblBodyCharset = New System.Windows.Forms.Label
        Me.cmbHeaderCharset = New System.Windows.Forms.ComboBox
        Me.lblHeaderCharset = New System.Windows.Forms.Label
        Me.groupBox_view = New System.Windows.Forms.GroupBox
        Me.chkBodyInTree = New System.Windows.Forms.CheckBox
        Me.chkHeaderInTree = New System.Windows.Forms.CheckBox
        Me.chkFieldsInTree = New System.Windows.Forms.CheckBox
        Me.panel1 = New System.Windows.Forms.Panel
        Me.groupBox_parser.SuspendLayout()
        Me.groupBox_view.SuspendLayout()
        Me.SuspendLayout()
        '
        ' groupBox_parser
        '
        Me.groupBox_parser.Controls.Add(Me.chk_UseBackgroundParser)
        Me.groupBox_parser.Controls.Add(Me.cmbActivatePartHandlers)
        Me.groupBox_parser.Controls.Add(Me.cmbBodyCharset)
        Me.groupBox_parser.Controls.Add(Me.lblBodyCharset)
        Me.groupBox_parser.Controls.Add(Me.cmbHeaderCharset)
        Me.groupBox_parser.Controls.Add(Me.lblHeaderCharset)
        Me.groupBox_parser.Dock = System.Windows.Forms.DockStyle.Top
        Me.groupBox_parser.Location = New System.Drawing.Point(0, 0)
        Me.groupBox_parser.Name = "groupBox_parser"
        Me.groupBox_parser.Size = New System.Drawing.Size(870, 160)
        Me.groupBox_parser.TabIndex = 0
        Me.groupBox_parser.TabStop = False
        Me.groupBox_parser.Text = " Message Parser Options "
        '
        ' chk_UseBackgroundParser
        '
        Me.chk_UseBackgroundParser.Location = New System.Drawing.Point(12, 124)
        Me.chk_UseBackgroundParser.Name = "chk_UseBackgroundParser"
        Me.chk_UseBackgroundParser.Size = New System.Drawing.Size(372, 24)
        Me.chk_UseBackgroundParser.TabIndex = 5
        Me.chk_UseBackgroundParser.Text = "Parsing Message in paralel thread"
        '
        ' cmbActivatePartHandlers
        '
        Me.cmbActivatePartHandlers.Location = New System.Drawing.Point(12, 92)
        Me.cmbActivatePartHandlers.Name = "cmbActivatePartHandlers"
        Me.cmbActivatePartHandlers.Size = New System.Drawing.Size(372, 24)
        Me.cmbActivatePartHandlers.TabIndex = 4
        Me.cmbActivatePartHandlers.Text = "Activate Part Handlers when message Loading"
        '
        ' cmbBodyCharset
        '
        Me.cmbBodyCharset.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
        Me.cmbBodyCharset.Location = New System.Drawing.Point(212, 56)
        Me.cmbBodyCharset.Name = "cmbBodyCharset"
        Me.cmbBodyCharset.Size = New System.Drawing.Size(176, 21)
        Me.cmbBodyCharset.TabIndex = 3
        '
        ' lblBodyCharset
        '
        Me.lblBodyCharset.Location = New System.Drawing.Point(212, 28)
        Me.lblBodyCharset.Name = "lblBodyCharset"
        Me.lblBodyCharset.Size = New System.Drawing.Size(176, 23)
        Me.lblBodyCharset.TabIndex = 2
        Me.lblBodyCharset.Text = "Default Body Charset"
        '
        ' cmbHeaderCharset
        '
        Me.cmbHeaderCharset.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
        Me.cmbHeaderCharset.Location = New System.Drawing.Point(12, 56)
        Me.cmbHeaderCharset.Name = "cmbHeaderCharset"
        Me.cmbHeaderCharset.Size = New System.Drawing.Size(180, 21)
        Me.cmbHeaderCharset.TabIndex = 1
        '
        ' lblHeaderCharset
        '
        Me.lblHeaderCharset.Location = New System.Drawing.Point(12, 28)
        Me.lblHeaderCharset.Name = "lblHeaderCharset"
        Me.lblHeaderCharset.Size = New System.Drawing.Size(180, 23)
        Me.lblHeaderCharset.TabIndex = 0
        Me.lblHeaderCharset.Text = "Default Header Charset"
        '
        ' groupBox_view
        '
        Me.groupBox_view.Controls.Add(Me.chkBodyInTree)
        Me.groupBox_view.Controls.Add(Me.chkHeaderInTree)
        Me.groupBox_view.Controls.Add(Me.chkFieldsInTree)
        Me.groupBox_view.Dock = System.Windows.Forms.DockStyle.Top
        Me.groupBox_view.Location = New System.Drawing.Point(0, 168)
        Me.groupBox_view.Name = "groupBox_view"
        Me.groupBox_view.Size = New System.Drawing.Size(870, 116)
        Me.groupBox_view.TabIndex = 2
        Me.groupBox_view.TabStop = False
        Me.groupBox_view.Text = " View Options for parsed messages   "
        '
        ' chkBodyInTree
        '
        Me.chkBodyInTree.Location = New System.Drawing.Point(12, 84)
        Me.chkBodyInTree.Name = "chkBodyInTree"
        Me.chkBodyInTree.Size = New System.Drawing.Size(372, 24)
        Me.chkBodyInTree.TabIndex = 2
        Me.chkBodyInTree.Text = "Body In Tree"
        '
        ' chkHeaderInTree
        '
        Me.chkHeaderInTree.Location = New System.Drawing.Point(12, 52)
        Me.chkHeaderInTree.Name = "chkHeaderInTree"
        Me.chkHeaderInTree.Size = New System.Drawing.Size(372, 24)
        Me.chkHeaderInTree.TabIndex = 1
        Me.chkHeaderInTree.Text = "Header In Tree"
        '
        ' chkFieldsInTree
        '
        Me.chkFieldsInTree.Location = New System.Drawing.Point(12, 20)
        Me.chkFieldsInTree.Name = "chkFieldsInTree"
        Me.chkFieldsInTree.Size = New System.Drawing.Size(372, 24)
        Me.chkFieldsInTree.TabIndex = 0
        Me.chkFieldsInTree.Text = "Fields In Tree"
        '
        ' panel1
        '
        Me.panel1.Dock = System.Windows.Forms.DockStyle.Top
        Me.panel1.Location = New System.Drawing.Point(0, 160)
        Me.panel1.Name = "panel1"
        Me.panel1.Size = New System.Drawing.Size(870, 8)
        Me.panel1.TabIndex = 1
        '
        ' MimeViewer_Options
        '
        Me.Controls.Add(groupBox_view)
        Me.Controls.Add(panel1)
        Me.Controls.Add(groupBox_parser)
        Me.Name = "MimeViewer_Options"
        Me.Size = New System.Drawing.Size(870, 434)
        Me.groupBox_parser.ResumeLayout(False)
        Me.groupBox_view.ResumeLayout(False)
        Me.ResumeLayout(False)

    End Sub 'InitializeComponent 
#End Region

    Private Shared Sub EnumCharsets(ByVal Category As String, ByVal Description As String, ByVal Name As String, ByVal Aliases As String, ByVal UserData As Object, ByRef [Stop] As Boolean)
        Dim myData As MimeViewer_Options
        Try
            myData = CType(UserData, MimeViewer_Options)
        Catch
            myData = Nothing
        End Try
        If Not (myData Is Nothing) Then
            myData.cmbHeaderCharset.Items.Add(Name)
            myData.cmbBodyCharset.Items.Add(Name)
        End If
    End Sub

    Friend fLocked As Boolean = True

    Private Sub MimeViewer_Options_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
        fCaption = "MIME Options"

        fLocked = True

        ' Fill Combo charsets names:
        Dim enumCharsets As New SBChSConv.TEnumCharsetsProc(AddressOf MimeViewer_Options.EnumCharsets)

        cmbHeaderCharset.BeginUpdate()
        Try
            cmbHeaderCharset.Items.Add("")
            cmbBodyCharset.Items.Add("")

            SBChSConv.Unit.EnumCharsets(enumCharsets, Me)
            cmbHeaderCharset.Sorted = True
            cmbBodyCharset.Sorted = True

        Finally
            cmbHeaderCharset.EndUpdate()
        End Try

        ' Fill other ProjectOptions:
        cmbActivatePartHandlers.Checked = ProjectOptions.fDefaultActivatePartHandlers
        chk_UseBackgroundParser.Checked = ProjectOptions.fUseBackgroundParser

        chkFieldsInTree.Checked = ProjectOptions.fFieldsInTree
        chkHeaderInTree.Checked = ProjectOptions.fHeaderInTree
        chkBodyInTree.Checked = ProjectOptions.fBodyInTree

        fLocked = False

    End Sub

    Protected Overrides Sub Init(ByVal messagePart As TElMessagePart, ByVal tagInfo As TagInfo, ByVal treeNodeItem As TreeNodeInfo, ByVal bShow As Boolean)
        MyBase.Init(messagePart, tagInfo, treeNodeItem, bShow)
        If treeNodeItem Is Nothing Then
            Return
        End If
    End Sub

    Private Sub cmbHeaderCharset_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles cmbHeaderCharset.SelectedIndexChanged
        If fLocked Then
            Return
        End If
        ProjectOptions.fDefaultHeaderCharset = cmbHeaderCharset.Text
    End Sub

    Private Sub cmbBodyCharset_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles cmbBodyCharset.SelectedIndexChanged
        If fLocked Then
            Return
        End If
        ProjectOptions.fDefaultHeaderCharset = cmbBodyCharset.Text
    End Sub

    Private Sub cmbActivatePartHandlers_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles cmbActivatePartHandlers.CheckedChanged
        If fLocked Then
            Return
        End If
        ProjectOptions.fDefaultActivatePartHandlers = cmbActivatePartHandlers.Checked
    End Sub

    Private Sub chk_UseBackgroundParser_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles chk_UseBackgroundParser.CheckedChanged
        If fLocked Then
            Return
        End If
        ProjectOptions.fUseBackgroundParser = chk_UseBackgroundParser.Checked
    End Sub


    Private Sub chkFieldsInTree_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles chkFieldsInTree.CheckedChanged
        If fLocked Then
            Return
        End If
        ProjectOptions.fFieldsInTree = chkFieldsInTree.Checked
    End Sub

    Private Sub chkHeaderInTree_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles chkHeaderInTree.CheckedChanged
        If fLocked Then
            Return
        End If
        ProjectOptions.fHeaderInTree = chkHeaderInTree.Checked
    End Sub

    Private Sub chkBodyInTree_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles chkBodyInTree.CheckedChanged
        If fLocked Then
            Return
        End If
        ProjectOptions.fBodyInTree = chkBodyInTree.Checked

    End Sub
End Class

Public NotInheritable Class ProjectOptions
    Public Shared fDefaultHeaderCharset As String = ""
    Public Shared fDefaultBodyCharset As String = ""
    Public Shared fDefaultActivatePartHandlers As Boolean = False
    Public Shared fUseBackgroundParser As Boolean = False

    Public Shared fParamsInTree As Boolean = True
    Public Shared fFieldsInTree As Boolean = True
    Public Shared fHeaderInTree As Boolean = True
    Public Shared fBodyInTree As Boolean = True
End Class

File: Contrib/SecureBlackbox.NET/Samples/C#/MIMEBlackbox/Viewer/MimeViewer_Options.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using System.Threading;

using SBMIME;
using SBChSConv;

namespace MimeViewer
{
	public class MimeViewer_Options : MimeViewer.MimeViewer_PlugControl
	{
     private System.Windows.Forms.GroupBox groupBox_parser;
     private System.Windows.Forms.Label label_HeaderCharset;
     private System.Windows.Forms.ComboBox comboBox_HeaderCharset;
     private System.Windows.Forms.Label label_BodyCharset;
     private System.Windows.Forms.ComboBox comboBox_BodyCharset;
     private System.Windows.Forms.CheckBox checkBox_ActivatePartHandlers;
     private System.Windows.Forms.CheckBox checkBox_UseBackgroundParser;
     private System.Windows.Forms.GroupBox groupBox_view;
     private System.Windows.Forms.CheckBox checkBox_BodyInTree;
     private System.Windows.Forms.CheckBox checkBox_HeaderInTree;
     private System.Windows.Forms.CheckBox checkBox_FieldsInTree;
     private System.Windows.Forms.Panel panel1;
		private System.ComponentModel.IContainer components = null;

		public MimeViewer_Options()
		{
			// This call is required by the Windows Form Designer.
			InitializeComponent();

			// TODO: Add any initialization after the InitializeComponent call
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			this.groupBox_parser = new System.Windows.Forms.GroupBox();
			this.checkBox_UseBackgroundParser = new System.Windows.Forms.CheckBox();
			this.checkBox_ActivatePartHandlers = new System.Windows.Forms.CheckBox();
			this.comboBox_BodyCharset = new System.Windows.Forms.ComboBox();
			this.label_BodyCharset = new System.Windows.Forms.Label();
			this.comboBox_HeaderCharset = new System.Windows.Forms.ComboBox();
			this.label_HeaderCharset = new System.Windows.Forms.Label();
			this.groupBox_view = new System.Windows.Forms.GroupBox();
			this.checkBox_BodyInTree = new System.Windows.Forms.CheckBox();
			this.checkBox_HeaderInTree = new System.Windows.Forms.CheckBox();
			this.checkBox_FieldsInTree = new System.Windows.Forms.CheckBox();
			this.panel1 = new System.Windows.Forms.Panel();
			this.groupBox_parser.SuspendLayout();
			this.groupBox_view.SuspendLayout();
			this.SuspendLayout();
			// 
			// groupBox_parser
			// 
			this.groupBox_parser.Controls.Add(this.checkBox_UseBackgroundParser);
			this.groupBox_parser.Controls.Add(this.checkBox_ActivatePartHandlers);
			this.groupBox_parser.Controls.Add(this.comboBox_BodyCharset);
			this.groupBox_parser.Controls.Add(this.label_BodyCharset);
			this.groupBox_parser.Controls.Add(this.comboBox_HeaderCharset);
			this.groupBox_parser.Controls.Add(this.label_HeaderCharset);
			this.groupBox_parser.Dock = System.Windows.Forms.DockStyle.Top;
			this.groupBox_parser.Location = new System.Drawing.Point(0, 0);
			this.groupBox_parser.Name = "groupBox_parser";
			this.groupBox_parser.Size = new System.Drawing.Size(877, 160);
			this.groupBox_parser.TabIndex = 0;
			this.groupBox_parser.TabStop = false;
			this.groupBox_parser.Text = " Message Parser Options ";
			// 
			// checkBox_UseBackgroundParser
			// 
			this.checkBox_UseBackgroundParser.Location = new System.Drawing.Point(12, 124);
			this.checkBox_UseBackgroundParser.Name = "checkBox_UseBackgroundParser";
			this.checkBox_UseBackgroundParser.Size = new System.Drawing.Size(372, 24);
			this.checkBox_UseBackgroundParser.TabIndex = 5;
			this.checkBox_UseBackgroundParser.Text = "Parse message in background";
			this.checkBox_UseBackgroundParser.CheckedChanged += new System.EventHandler(this.checkBox_UseBackgroundParser_CheckedChanged);
			// 
			// checkBox_ActivatePartHandlers
			// 
			this.checkBox_ActivatePartHandlers.Location = new System.Drawing.Point(12, 92);
			this.checkBox_ActivatePartHandlers.Name = "checkBox_ActivatePartHandlers";
			this.checkBox_ActivatePartHandlers.Size = new System.Drawing.Size(372, 24);
			this.checkBox_ActivatePartHandlers.TabIndex = 4;
			this.checkBox_ActivatePartHandlers.Text = "Activate Part Handlers when loading a message";
			this.checkBox_ActivatePartHandlers.CheckedChanged += new System.EventHandler(this.checkBox_ActivatePartHandlers_CheckedChanged);
			// 
			// comboBox_BodyCharset
			// 
			this.comboBox_BodyCharset.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
			this.comboBox_BodyCharset.Location = new System.Drawing.Point(212, 56);
			this.comboBox_BodyCharset.Name = "comboBox_BodyCharset";
			this.comboBox_BodyCharset.Size = new System.Drawing.Size(176, 21);
			this.comboBox_BodyCharset.TabIndex = 3;
			this.comboBox_BodyCharset.SelectedIndexChanged += new System.EventHandler(this.comboBox_BodyCharset_SelectedIndexChanged);
			// 
			// label_BodyCharset
			// 
			this.label_BodyCharset.Location = new System.Drawing.Point(212, 28);
			this.label_BodyCharset.Name = "label_BodyCharset";
			this.label_BodyCharset.Size = new System.Drawing.Size(176, 23);
			this.label_BodyCharset.TabIndex = 2;
			this.label_BodyCharset.Text = "Default Body Charset";
			// 
			// comboBox_HeaderCharset
			// 
			this.comboBox_HeaderCharset.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
			this.comboBox_HeaderCharset.Location = new System.Drawing.Point(12, 56);
			this.comboBox_HeaderCharset.Name = "comboBox_HeaderCharset";
			this.comboBox_HeaderCharset.Size = new System.Drawing.Size(180, 21);
			this.comboBox_HeaderCharset.TabIndex = 1;
			this.comboBox_HeaderCharset.SelectedIndexChanged += new System.EventHandler(this.comboBox_HeaderCharset_SelectedIndexChanged);
			// 
			// label_HeaderCharset
			// 
			this.label_HeaderCharset.Location = new System.Drawing.Point(12, 28);
			this.label_HeaderCharset.Name = "label_HeaderCharset";
			this.label_HeaderCharset.Size = new System.Drawing.Size(180, 23);
			this.label_HeaderCharset.TabIndex = 0;
			this.label_HeaderCharset.Text = "Default Header Charset";
			// 
			// groupBox_view
			// 
			this.groupBox_view.Controls.Add(this.checkBox_BodyInTree);
			this.groupBox_view.Controls.Add(this.checkBox_HeaderInTree);
			this.groupBox_view.Controls.Add(this.checkBox_FieldsInTree);
			this.groupBox_view.Dock = System.Windows.Forms.DockStyle.Top;
			this.groupBox_view.Location = new System.Drawing.Point(0, 168);
			this.groupBox_view.Name = "groupBox_view";
			this.groupBox_view.Size = new System.Drawing.Size(877, 116);
			this.groupBox_view.TabIndex = 2;
			this.groupBox_view.TabStop = false;
			this.groupBox_view.Text = " View Options for parsed messages   ";
			// 
			// checkBox_BodyInTree
			// 
			this.checkBox_BodyInTree.Location = new System.Drawing.Point(12, 84);
			this.checkBox_BodyInTree.Name = "checkBox_BodyInTree";
			this.checkBox_BodyInTree.Size = new System.Drawing.Size(372, 24);
			this.checkBox_BodyInTree.TabIndex = 2;
			this.checkBox_BodyInTree.Text = "Body In Tree";
			this.checkBox_BodyInTree.CheckedChanged += new System.EventHandler(this.checkBox_BodyInTree_CheckedChanged);
			// 
			// checkBox_HeaderInTree
			// 
			this.checkBox_HeaderInTree.Location = new System.Drawing.Point(12, 52);
			this.checkBox_HeaderInTree.Name = "checkBox_HeaderInTree";
			this.checkBox_HeaderInTree.Size = new System.Drawing.Size(372, 24);
			this.checkBox_HeaderInTree.TabIndex = 1;
			this.checkBox_HeaderInTree.Text = "Header In Tree";
			this.checkBox_HeaderInTree.CheckedChanged += new System.EventHandler(this.checkBox_HeaderInTree_CheckedChanged);
			// 
			// checkBox_FieldsInTree
			// 
			this.checkBox_FieldsInTree.Location = new System.Drawing.Point(12, 20);
			this.checkBox_FieldsInTree.Name = "checkBox_FieldsInTree";
			this.checkBox_FieldsInTree.Size = new System.Drawing.Size(372, 24);
			this.checkBox_FieldsInTree.TabIndex = 0;
			this.checkBox_FieldsInTree.Text = "Fields In Tree";
			this.checkBox_FieldsInTree.CheckedChanged += new System.EventHandler(this.checkBox_FieldsInTree_CheckedChanged);
			// 
			// panel1
			// 
			this.panel1.Dock = System.Windows.Forms.DockStyle.Top;
			this.panel1.Location = new System.Drawing.Point(0, 160);
			this.panel1.Name = "panel1";
			this.panel1.Size = new System.Drawing.Size(877, 8);
			this.panel1.TabIndex = 1;
			// 
			// MimeViewer_Options
			// 
			this.Controls.Add(this.groupBox_view);
			this.Controls.Add(this.panel1);
			this.Controls.Add(this.groupBox_parser);
			this.Name = "MimeViewer_Options";
			this.Size = new System.Drawing.Size(877, 656);
			this.Load += new System.EventHandler(this.MimeViewer_Options_Load);
			this.groupBox_parser.ResumeLayout(false);
			this.groupBox_view.ResumeLayout(false);
			this.ResumeLayout(false);

		}
		#endregion

     private static void EnumCharsets(string Category, string Description, string Name,
       string Aliases, object UserData, ref bool Stop)
     {
       MimeViewer_Options myData = UserData as MimeViewer_Options;
       if (myData != null){
         myData.comboBox_HeaderCharset.Items.Add(Name);
         myData.comboBox_BodyCharset.Items.Add(Name);
       }
     }

     internal bool fLocked = true;

     private void MimeViewer_Options_Load(object sender, System.EventArgs e)
     {

       fCaption = "MIME Options";

       fLocked = true;

       // Fill Combo charsets names:

       SBChSConv.TEnumCharsetsProc enumCharsets = // create delegate (EnumCharsets):
         new SBChSConv.TEnumCharsetsProc(MimeViewer_Options.EnumCharsets);

       comboBox_HeaderCharset.BeginUpdate();
       try
       {
         comboBox_HeaderCharset.Items.Add("");
         comboBox_BodyCharset.Items.Add("");

         SBChSConv.Unit.EnumCharsets(enumCharsets, this);
         comboBox_HeaderCharset.Sorted = true;
         comboBox_BodyCharset.Sorted = true;

       }
       finally
       {
         comboBox_HeaderCharset.EndUpdate();
       }

       // Fill other ProjectOptions:

       checkBox_ActivatePartHandlers.Checked = ProjectOptions.fDefaultActivatePartHandlers;
       checkBox_UseBackgroundParser.Checked = ProjectOptions.fUseBackgroundParser;

       checkBox_FieldsInTree.Checked = ProjectOptions.fFieldsInTree;
       checkBox_HeaderInTree.Checked = ProjectOptions.fHeaderInTree;
       checkBox_BodyInTree.Checked = ProjectOptions.fBodyInTree;

       fLocked = false;
     }

     protected override void Init(TElMessagePart messagePart, TagInfo tagInfo, TreeNodeInfo treeNodeItem, bool bShow)
     {
       base.Init(messagePart, tagInfo, treeNodeItem, bShow);
       if (treeNodeItem == null)
         return;
     }

     private void comboBox_HeaderCharset_SelectedIndexChanged(object sender, System.EventArgs e)
     {
       if ( fLocked ) return;
       ProjectOptions.fDefaultHeaderCharset = comboBox_HeaderCharset.Text;
     }

     private void comboBox_BodyCharset_SelectedIndexChanged(object sender, System.EventArgs e)
     {
       if ( fLocked ) return;
       ProjectOptions.fDefaultHeaderCharset = comboBox_BodyCharset.Text;
     }

     private void checkBox_ActivatePartHandlers_CheckedChanged(object sender, System.EventArgs e)
     {
       if ( fLocked ) return;
       ProjectOptions.fDefaultActivatePartHandlers = checkBox_ActivatePartHandlers.Checked;
     }

     private void checkBox_UseBackgroundParser_CheckedChanged(object sender, System.EventArgs e)
     {
       if ( fLocked ) return;
       ProjectOptions.fUseBackgroundParser = checkBox_UseBackgroundParser.Checked;
     }

     private void checkBox_FieldsInTree_CheckedChanged(object sender, System.EventArgs e)
     {
       if ( fLocked ) return;
       ProjectOptions.fFieldsInTree = checkBox_FieldsInTree.Checked;
     }

     private void checkBox_HeaderInTree_CheckedChanged(object sender, System.EventArgs e)
     {
       if ( fLocked ) return;
       ProjectOptions.fHeaderInTree = checkBox_HeaderInTree.Checked;
     }

     private void checkBox_BodyInTree_CheckedChanged(object sender, System.EventArgs e)
     {
       if ( fLocked ) return;
       ProjectOptions.fBodyInTree = checkBox_BodyInTree.Checked;
     }

	}//of: MimeViewer_Options

	sealed public class ProjectOptions
	{
      public static string fDefaultHeaderCharset = "";
      public static string fDefaultBodyCharset = "";
      public static bool fDefaultActivatePartHandlers = false;
      public static bool fUseBackgroundParser = false;

      public static bool fParamsInTree = true;
      public static bool fFieldsInTree = true;
      public static bool fHeaderInTree = true;
      public static bool fBodyInTree = true;

	}//of: ProjectOptions
}
